src/WorldIDIdentityManagerImplV1.sol

```diff
@@ -782,17 +782,102 @@ contract WorldIDIdentityManagerImplV1 is WorldIDImpl, IWorldID {
         view
         virtual
     {
-        bool previousIsZero = false;
-
-        for (uint256 i = 0; i < identityCommitments.length; ++i) {
-            uint256 commitment = identityCommitments[i];
-            if (previousIsZero && commitment != EMPTY_LEAF) {
-                revert InvalidCommitment(i);
+        // this will be assigned an unreduced element if one is found
+        // note that zero is never unreduced, so it's a safe no-revert state
+        uint256 revertOnUnreduced = 0;
+        // this will be assigned a non-zero element index that occurs after a zero element
+        // note that this definition means an actual revert always happens at index at least 1,
+        // so zero is a safe no-revert state
+        uint256 revertOnInvalid = 0;
+        assembly {
```

Comment:
Does assembly lead to meaningful savings compared to the (substantial) code complexity it introduces? Auditors are going to call us out on this for sure. Can we compare against a simple:

```solidity
        bool previousIsZero = false;
        for (uint256 i = 0; i < identityCommitments.length; ++i) {
            uint256 commitment = identityCommitments[i];
            if (previousIsZero && commitment != EMPTY_LEAF) {
                revert InvalidCommitment(i);
            }
            if (commitment >= SNARK_SCALAR_FIELD) {
                revert UnreducedElement(UnreducedElementType.IdentityCommitment, commitment);
            previousIsZero = commitment == EMPTY_LEAF;
        }
```